### DISCUSSION TITLE
npm lock file out of sync

### DISCUSSION THREAD

**Discussion by json-derulo** [renovatebot/renovate#31984] [Request Help] [answered]
### How are you running Renovate?

A Mend.io-hosted app

### Please tell us more about your question or problem

I am using Renovate with lock file maintenance on many Angular projects (npm). I never had a problem with it, until this week. During lock file maintenance, Renovate updated the `package-lock.json` file. However when running `npm ci` on the project with the updated lockfile, I am getting the following error:

```
npm error code EUSAGE
npm error
npm error `npm ci` can only install packages when your package.json and package-lock.json or npm-shrinkwrap.json are in sync. Please update your lock file with `npm install` before continuing.
npm error
npm error Invalid: lock file's chokidar@4.0.1 does not satisfy chokidar@3.6.0
npm error Missing: chokidar@4.0.1 from lock file
npm error Missing: glob-parent@5.1.2 from lock file
npm error Invalid: lock file's readdirp@4.0.2 does not satisfy readdirp@3.6.0
npm error Missing: picomatch@2.3.1 from lock file
npm error Missing: readdirp@4.0.2 from lock file
```

It seems like the generated lock file is invalid, some major versions are updated even if some packages don't allow it. This only happens on my Angular projects, other projects seem fine. Here is an example PR: https://github.com/json-derulo/angular-ecmascript-intl/pull/469

### Logs (if relevant)

<details><summary>Logs</summary>

```

Replace this text with your logs, between the starting and ending triple backticks

```

</details>

**Comment by rarkins** [comment]
I was able to reproduce this by:
- Cloning your repo
- Checking out this branch/PR
- Deleting `package-lock.json`
- Ensuring that `node_modules` was missing/empty
- Running `npm i --package-lock-only`

The resulting `package-lock.json` was the same.

In conclusion, `npm i --package-lock-only` is buggy, so Renovate would need to run `npm i` instead, which is slower but should always produce a correct result.

I wonder what it is about this repo/dependency list which triggers it. e.g. some uncommon case or peer dependencies or something like that.

**Reply by rarkins** [reply, marked as answer]
Unfortunately the reproduction turns out to be even simpler:

- Clone the repo
- Run `npm i`
- Run `npm ci`

Strangely, running `npm i` a second time seems to fix the problem. See log below:

```
/tmp/clone/angular-ecmascript-intl main
❯ rm package-lock.json 
Execution time: 0.03s

/tmp/clone/angular-ecmascript-intl main*
❯ npm i
npm warn deprecated inflight@1.0.6: This module is not supported, and leaks memory. Do not use it. Check out lru-cache if you want a good and tested way to coalesce async requests by a key value, which is much more comprehensive and powerful.
npm warn deprecated rimraf@3.0.2: Rimraf versions prior to v4 are no longer supported
npm warn deprecated glob@7.2.3: Glob versions prior to v9 are no longer supported

added 1230 packages, and audited 1231 packages in 13s

204 packages are looking for funding
  run `npm fund` for details

found 0 vulnerabilities
Execution time: 12.82s

/tmp/clone/angular-ecmascript-intl main* 13s
❯ npm ci
npm error code EUSAGE
npm error
npm error `npm ci` can only install packages when your package.json and package-lock.json or npm-shrinkwrap.json are in sync. Please update your lock file with `npm install` before continuing.
npm error
npm error Invalid: lock file's chokidar@4.0.1 does not satisfy chokidar@3.6.0
npm error Missing: chokidar@4.0.1 from lock file
npm error Missing: glob-parent@5.1.2 from lock file
npm error Invalid: lock file's readdirp@4.0.2 does not satisfy readdirp@3.6.0
npm error Missing: picomatch@2.3.1 from lock file
npm error Missing: readdirp@4.0.2 from lock file
npm error
npm error Clean install a project
npm error
npm error Usage:
npm error npm ci
npm error
npm error Options:
npm error [--install-strategy <hoisted|nested|shallow|linked>] [--legacy-bundling]
npm error [--global-style] [--omit <dev|optional|peer> [--omit <dev|optional|peer> ...]]
npm error [--include <prod|dev|optional|peer> [--include <prod|dev|optional|peer> ...]]
npm error [--strict-peer-deps] [--foreground-scripts] [--ignore-scripts] [--no-audit]
npm error [--no-bin-links] [--no-fund] [--dry-run]
npm error [-w|--workspace <workspace-name> [-w|--workspace <workspace-name> ...]]
npm error [-ws|--workspaces] [--include-workspace-root] [--install-links]
npm error
npm error aliases: clean-install, ic, install-clean, isntall-clean
npm error
npm error Run "npm help ci" for more info
npm error A complete log of this run can be found in: /Users/rhys/.npm/_logs/2024-10-16T15_45_45_170Z-debug-0.log
Execution time: 0.57s

/tmp/clone/angular-ecmascript-intl main*
❯ npm i

added 5 packages, removed 16 packages, changed 2 packages, and audited 1220 packages in 1s

204 packages are looking for funding
  run `npm fund` for details

found 0 vulnerabilities
Execution time: 1.15s

/tmp/clone/angular-ecmascript-intl main*
❯ npm ci
npm warn deprecated inflight@1.0.6: This module is not supported, and leaks memory. Do not use it. Check out lru-cache if you want a good and tested way to coalesce async requests by a key value, which is much more comprehensive and powerful.
npm warn deprecated rimraf@3.0.2: Rimraf versions prior to v4 are no longer supported
npm warn deprecated glob@7.2.3: Glob versions prior to v9 are no longer supported

added 1219 packages, and audited 1220 packages in 11s

204 packages are looking for funding
  run `npm fund` for details

found 0 vulnerabilities
Execution time: 10.88s

```

Definitely seems like an `npm` bug.

**Reply by json-derulo** [reply]
Thanks a lot for your thorough investigation! I have created an issue in the npm repo: https://github.com/npm/cli/issues/7841

**Reply by GKersten** [reply]
Running into the same issue and found the issue on NPM repo. Is there a known work-around for renovate in the meantime?

**Reply by rarkins** [reply]
If you're self-hosted, you could add another `npm install` to `postUpgradeTasks`

**Reply by rarkins** [reply]
Someone suggested a workaround in https://github.com/npm/cli/issues/6787

**Reply by GKersten** [reply]
[this](https://github.com/npm/cli/issues/6787#issuecomment-2432828489) is the work around you are talking about, that could indeed help.

The `npm install` in `postUpgradeTasks` is what I tried, but than I ran into this https://github.com/renovatebot/renovate/discussions/14446 because we are using a private NPM feed.

**Comment by patrickuhlmann** [comment]
Since two weeks there is a section about this problem in the official documentation: https://docs.renovatebot.com/modules/manager/npm/#invalid-lock-file-npm-ci-fails

A new option `"postUpdateOptions": ["npmInstallTwice"]` has been introduced.

**Reply by grantaveryatgfs** [reply]
Hi, I've been running into this issue recently, and added the `"postUpdateOptions": ["npmInstallTwice"]` rule to my repo's `renovate.json` config file. However, despite seeing Renovate Runner printing the expected debug logs (see https://github.com/renovatebot/renovate/blob/main/lib/modules/manager/npm/post-update/npm.ts#L199-L200) in the runner job, the MR it created was still attempting to remove necessary npm lockfile dependencies (for `@rollup`, in my case).

I'll also note that this issue has been occurring for every single Renovate MR for the past week or so, all for the same `@rollup` lockfile dependencies. This seems to run counter to what the Renovate docs say about "npm itself _sometimes_ generates invalid lock files which fail npm ci".

Has anyone else run into this repeated issue before, or seen `npmInstallTwice` not solve it, and are there any ideas for how to work around it more directly?

<details>
<summary><b>Excerpt of my Renovate Runner logs, with "Running npm install twice" and two "Executing command ... npm install"</b></summary>

```
DEBUG: Spawning npm install to create apps/web-chat-ui/package-lock.json (repository=gfs/products/customer-sales/gordon-now/gordon-now, branch=renovate/uuid-13.x)
DEBUG: Found npm constraint in package.json engines: ^11.0.0 (repository=gfs/products/customer-sales/gordon-now/gordon-now, branch=renovate/uuid-13.x)
DEBUG: Updating lock file only (repository=gfs/products/customer-sales/gordon-now/gordon-now, branch=renovate/uuid-13.x)
DEBUG: No node constraint found - using latest (repository=gfs/products/customer-sales/gordon-now/gordon-now, branch=renovate/uuid-13.x)
DEBUG: Running npm install twice (repository=gfs/products/customer-sales/gordon-now/gordon-now, branch=renovate/uuid-13.x)
DEBUG: Setting CONTAINERBASE_CACHE_DIR to /builds/gfs/products/customer-sales/gordon-now/renovate-runner/renovate/cache/containerbase (repository=gfs/products/customer-sales/gordon-now/gordon-now, branch=renovate/uuid-13.x)
DEBUG: Using containerbase dynamic installs (repository=gfs/products/customer-sales/gordon-now/gordon-now, branch=renovate/uuid-13.x)
DEBUG: Resolved stable matching version (repository=gfs/products/customer-sales/gordon-now/gordon-now, branch=renovate/uuid-13.x)
       "toolName": "node",
       "constraint": null,
       "resolvedVersion": "22.19.0"
DEBUG: Resolved stable matching version (repository=gfs/products/customer-sales/gordon-now/gordon-now, branch=renovate/uuid-13.x)
       "toolName": "npm",
       "constraint": "^11.0.0",
       "resolvedVersion": "11.6.0"
DEBUG: Executing command (repository=gfs/products/customer-sales/gordon-now/gordon-now, branch=renovate/uuid-13.x)
       "command": "install-tool node 22.19.0"
DEBUG: exec completed (repository=gfs/products/customer-sales/gordon-now/gordon-now, branch=renovate/uuid-13.x)
       "durationMs": 4260,
       "stdout": "[18:39:52.573] INFO (311): Installing tool node@22.19.0...\nv22.19.0\n10.9.3\n0.34.0\n[18:39:55.757] INFO (311): Install tool node succeeded in 3.1s.\n",
       "stderr": ""
DEBUG: Executing command (repository=gfs/products/customer-sales/gordon-now/gordon-now, branch=renovate/uuid-13.x)
       "command": "install-tool npm 11.6.0"
DEBUG: exec completed (repository=gfs/products/customer-sales/gordon-now/gordon-now, branch=renovate/uuid-13.x)
       "durationMs": 4270,
       "stdout": "[18:39:56.851] INFO (348): Installing npm npm@11.6.0...\n11.6.0\n[18:40:00.169] INFO (348): Install tool npm succeeded in 3.3s.\n",
       "stderr": "npm warn Unknown env config \"store\". This will stop working in the next major version of npm.\n"
DEBUG: Executing command (repository=gfs/products/customer-sales/gordon-now/gordon-now, branch=renovate/uuid-13.x)
       "command": "hash -d npm 2>/dev/null || true"
DEBUG: exec completed (repository=gfs/products/customer-sales/gordon-now/gordon-now, branch=renovate/uuid-13.x)
       "durationMs": 20,
       "stdout": "",
       "stderr": ""
DEBUG: Executing command (repository=gfs/products/customer-sales/gordon-now/gordon-now, branch=renovate/uuid-13.x)
       "command": "npm install --package-lock-only --no-audit --ignore-scripts"
DEBUG: exec completed (repository=gfs/products/customer-sales/gordon-now/gordon-now, branch=renovate/uuid-13.x)
       "durationMs": 1601,
       "stdout": "\nup to date in 1s\n",
       "stderr": "npm warn Unknown env config \"store\". This will stop working in the next major version of npm.\n"
DEBUG: Executing command (repository=gfs/products/customer-sales/gordon-now/gordon-now, branch=renovate/uuid-13.x)
       "command": "npm install --package-lock-only --no-audit --ignore-scripts"
DEBUG: exec completed (repository=gfs/products/customer-sales/gordon-now/gordon-now, branch=renovate/uuid-13.x)
       "durationMs": 2777,
       "stdout": "\nup to date in 3s\n",
       "stderr": "npm warn Unknown env config \"store\". This will stop working in the next major version of npm.\n"
DEBUG: apps/web-chat-ui/package-lock.json needs updating (repository=gfs/products/customer-sales/gordon-now/gordon-now, branch=renovate/uuid-13.x)
DEBUG: Updated 1 lock files (repository=gfs/products/customer-sales/gordon-now/gordon-now, branch=renovate/uuid-13.x)
       "updatedArtifacts": ["apps/web-chat-ui/package-lock.json"]
DEBUG: 2 file(s) to commit (repository=gfs/products/customer-sales/gordon-now/gordon-now, branch=renovate/uuid-13.x)
```

</details>

<details>
<summary><b>`npm ci` logs showing that the MR Renovate created was missing a number of dependencies for `@rollup` in the lockfile it updated </b></summary>

$ npm ci --cache .npm --prefer-offline
npm error code EUSAGE
npm error
npm error `npm ci` can only install packages when your package.json and package-lock.json or npm-shrinkwrap.json are in sync. Please update your lock file with `npm install` before continuing.
npm error
npm error Missing: @rollup/rollup-android-arm-eabi@4.46.2 from lock file
npm error Missing: @rollup/rollup-android-arm64@4.46.2 from lock file
npm error Missing: @rollup/rollup-darwin-arm64@4.46.2 from lock file
npm error Missing: @rollup/rollup-darwin-x64@4.46.2 from lock file
npm error Missing: @rollup/rollup-freebsd-arm64@4.46.2 from lock file
npm error Missing: @rollup/rollup-freebsd-x64@4.46.2 from lock file
npm error Missing: @rollup/rollup-linux-arm-gnueabihf@4.46.2 from lock file
npm error Missing: @rollup/rollup-linux-arm-musleabihf@4.46.2 from lock file
npm error Missing: @rollup/rollup-linux-arm64-gnu@4.46.2 from lock file
npm error Missing: @rollup/rollup-linux-arm64-musl@4.46.2 from lock file
npm error Missing: @rollup/rollup-linux-loongarch64-gnu@4.46.2 from lock file
npm error Missing: @rollup/rollup-linux-ppc64-gnu@4.46.2 from lock file
npm error Missing: @rollup/rollup-linux-riscv64-gnu@4.46.2 from lock file
npm error Missing: @rollup/rollup-linux-riscv64-musl@4.46.2 from lock file
npm error Missing: @rollup/rollup-linux-s390x-gnu@4.46.2 from lock file
npm error Missing: @rollup/rollup-linux-x64-gnu@4.46.2 from lock file
npm error Missing: @rollup/rollup-linux-x64-musl@4.46.2 from lock file
npm error Missing: @rollup/rollup-win32-arm64-msvc@4.46.2 from lock file
npm error Missing: @rollup/rollup-win32-ia32-msvc@4.46.2 from lock file
npm error Missing: @rollup/rollup-win32-x64-msvc@4.46.2 from lock file
npm error
npm error Clean install a project

</details>

**Reply by RahulGautamSingh** [reply]
Not sure what we could do to fix the issue atm as it's an npm bug. Some users, however, tried regenerating their repo lockfiles manually, which seemed to have prevented the error from occurring in future renovate PRs.

**Reply by patrickuhlmann** [reply]
We have seen the same in our projects. npmInstallTwice not solving it and @rollup mentioned as problem when npm ci is executed. At the moment we have not conducted a deeper analysis.

**Reply by gaeljw** [reply]
Same here: still got an error even though we use `npmInstallTwice`. Fixing manually on one PR didn't help the other PRs even after forcing Renovate to re-run. Each PR creates a "bad" `package-lock.json` file.

**Reply by gaeljw** [reply]
After upgrading some dependencies manually (Angular) and using `^` instead of a mix of `~` and pinned version, Renovate works back as normal. I'm not clear on what exactly fixed it.

**Reply by wallind** [reply]
> After upgrading some dependencies manually (Angular) ... works back as normal

same here. These versions for us:

<img width="446" height="248" alt="image" src="https://github.com/user-attachments/assets/b5a30e34-4078-409b-afb9-f582aee20464" />

**Reply by mhoritani** [reply]
Has anybody found a reliable way of fixing this? We have tried `npmInstallTwice`, as well as updating angular related dependencies manually. All PRs Renovate creates are faulty, deleting random bits from `package-lock.json` resulting in an error when running `npm ci`.

**Reply by sleidig** [reply]
`npmInstallTwice` hasn't solved the problem for us either :thinking:

**Reply by mhoritani** [reply]
As an update here, what eventually did the trick for us was:

- Upgrading to `npm@11`
- Recreating the lock file with `npm@11` (make sure to remove all node_modules directories before you do so)
- Forcing Renovate to use `npm@11 `as well

Seems like npm@11 treats optional dependencies differently and helped resolve this issue

**Reply by cypressSap** [reply]
`"postUpdateOptions": ["npmInstallTwice"]` helped, but Renovate was still removing lots of `"peer": true` lines from the lockfile that would be re-added by `npm install`.

`"constraints": { "npm": ">= 11" }` seems to be the missing piece to eliminate that churn, thanks @mhoritani!

**Comment by patrickuhlmann** [comment]
Did you also delete the whole package-lock.json? For me it started working again after I updated all dependencies to the latest version and also deleted the package-lock.json and node_modules folder and executing npm install once manually - effectively create a fully updated fresh package-lock.json
